### PR TITLE
fix: akschually skip existing related assets

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

Makes the related assets generation script behave as it should and *not* regenerate for assets which already have a `relatedAssetId`, matching the intent of this comment:

https://github.com/shapeshift/web/blob/b6365b48b4db6b79726696b29dcb554ddac33890/scripts/generateAssetData/generateRelatedAssetIndex/generateRelatedAssetIndex.ts#L172

Also brings a `rebuildAll` option if we ever want to do a full regen, including assets with an existing `relatedAssetKey`.

## Pull Request Type

- [x] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

- relates to https://github.com/shapeshift/web/issues/7303

Low - ensure assets and related assets still look sane

> What protocols, transaction types or contract interactions might be affected by this PR?

## Testing

- Ensure assets (including portals assets) still look sane
- Ensure related assets still look sane for major assets (e.g USDC)
- Ensure manual related assets still look sane (FOX on Arbitrum, WETH)

### Engineering

- ☝🏽 
- Ensure running `yarn generate:related-asset-index` completes almost instantly and doesn't produce any `relatedAssetIndex.json` diff (i.e`generateRelatedAssetIndex() done. Successes: 0, Failures: 0`)
- `yarn generate:asset-data` is now blazingly fast

### Operations

- ☝🏽 

## Screenshots (if applicable)

Tested against develop (right):

image
image
image

### CODE REVIEW

File: packages/types/src/base.ts [hunk 1]
```diff
@@ -57,7 +57,7 @@ export type Asset = {
   explorer: string
   explorerTxLink: string
   explorerAddressLink: string
-  relatedAssetKey?: AssetId
+  relatedAssetKey: AssetId | null | undefined
```
Comment (gomesalexandre) [inline packages/types/src/base.ts:60]: The crux of this PR - JSON doesn't support `undefined`.

By making the `relatedAssetKey` required and having it as maybe null or undefined, we ensure that it is either `undefined` (e.g base assets with known no related assets, or generated assets for which we don't know related assets yet, to be fetched by the `generateRelatedAssetIndex` script), or an explicit `null` in the JSON (i.e related asset indexes have been fetched, but there are none).

In other words, by having `AssetId | null` in the JSON, we are now able to not refetch existing related AssetIds with a simple `  if (existingRelatedAssetKey !== undefined) return` check

File: scripts/generateAssetData/generateRelatedAssetIndex/generateRelatedAssetIndex.ts [hunk 2]
```diff
@@ -166,13 +167,27 @@ let sadCount = 0
...
+  assetData: Record<AssetId, PartialFields<Asset, 'relatedAssetKey'>>,
   relatedAssetIndex: Record<AssetId, AssetId[]>,
+  throttle: () => Promise<void>,
 ): Promise<void> => {
-  // don't fetch if we've already got the data from a previous request
   const existingRelatedAssetKey = assetData[assetId].relatedAssetKey
-  if (existingRelatedAssetKey) {
-    return
+  // We already have an existing relatedAssetKey, so we don't need to fetch it again
+  if (existingRelatedAssetKey !== undefined) return
```
Comment (gomesalexandre) [inline scripts/generateAssetData/generateRelatedAssetIndex/generateRelatedAssetIndex.ts:176]: The second crux of this PR - as mentioned in https://github.com/shapeshift/web/pull/7408/files#r1688698060, `relatedAssetKey` will now *always* be included at generation time, effectively always `undefined` for now, and then set to `AssetId | undefined` by the generation script.

The next time it runs, if it encounters any `relatedAssetKey` which is not undefined, it means that we already know this asset and have successfully its related assets, so we can skip refetching.

File: scripts/generateAssetData/generateRelatedAssetIndex/generateRelatedAssetIndex.ts [hunk 3]
```diff
@@ -196,26 +211,29 @@ const processRelatedAssetIds = async (
...
+  if (hasRelatedAssets) {
+    // attach the relatedAssetKey for all related assets including the primary implementation
+    for (const relatedAssetId of mergedRelatedAssetIds) {
+      if (assetData[relatedAssetId]) {
+        assetData[relatedAssetId].relatedAssetKey = relatedAssetKey
+      }
+    }
+    relatedAssetIndex[relatedAssetKey] = mergedRelatedAssetIds
+  } else {
+    // If there are no related assets, set relatedAssetKey to null
```
Comment (gomesalexandre) [inline scripts/generateAssetData/generateRelatedAssetIndex/generateRelatedAssetIndex.ts:229]: The magic line

File: scripts/generateAssetData/generateRelatedAssetIndex/generateRelatedAssetIndex.ts [hunk 4]
```diff
@@ -196,26 +211,29 @@ const processRelatedAssetIds = async (
...
+    assetData[assetId].relatedAssetKey = null
   }
-  relatedAssetIndex[relatedAssetKey] = mergedRelatedAssetIds
-  return
-}
 
-export const generateRelatedAssetIndex = async () => {
-  console.log('generateRelatedAssetIndex() starting')
+  await throttle()
```
Comment (gomesalexandre) [inline scripts/generateAssetData/generateRelatedAssetIndex/generateRelatedAssetIndex.ts:233]: Note we now throttle at the end of this method - no need to throttle if we're skipping the current fetching run.

File: scripts/generateAssetData/generateRelatedAssetIndex/generateRelatedAssetIndex.ts [hunk 5]
```diff
@@ -229,11 +247,18 @@ export const generateRelatedAssetIndex = async () => {
   const generatedAssetData: AssetsById = JSON.parse(
     await fs.promises.readFile(generatedAssetsPath, 'utf8'),
   )
-  const relatedAssetIndex: Record<AssetId, AssetId[]> = {}
-  const assetDataWithRelatedAssetKeys: AssetsById = { ...generatedAssetData }
+  const relatedAssetIndex: Record<AssetId, AssetId[]> = JSON.parse(
```
Comment (gomesalexandre) [inline scripts/generateAssetData/generateRelatedAssetIndex/generateRelatedAssetIndex.ts:250]: The other crux of this PR - now that we skip generation for known assets (unless `rebuildAll` is explicitly set to true), we can't just initialize this to an empty object, else we'd end up with an empty `relatedAssetIndex.json` as a result.

File: scripts/generateAssetData/utils/portals.ts [hunk 6]
```diff
@@ -138,6 +138,7 @@ export const getPortalTokens = async (nativeAsset: Asset): Promise<Asset[]> => {
       symbol: token.symbol,
       chainId: nativeAsset.chainId,
       assetId,
+      relatedAssetKey: undefined,
```
Comment (gomesalexandre) [inline scripts/generateAssetData/utils/portals.ts:141]: `undefined` vs. null as we still want to fetch related assets afterwards

File: scripts/generateAssetData/generateAssetData.ts [hunk 7]
```diff
@@ -90,7 +95,17 @@ const generateAssetData = async () => {
       .includes(asset.name)
   }
 
+  const currentGeneratedAssetData: AssetsByIdPartial = JSON.parse(
+    await fs.promises.readFile(generatedAssetsPath, 'utf8'),
+  )
+
   const generatedAssetData = orderedAssetList.reduce<AssetsById>((acc, asset) => {
+    const currentGeneratedAssetId = currentGeneratedAssetData[asset.assetId]
+    // Ensures we don't overwrite existing relatedAssetIndex with the generated one, triggering a refetch
```
Comment (gomesalexandre) [inline scripts/generateAssetData/generateAssetData.ts:104]: Pretty much what it says on the comment

File: scripts/generateAssetData/generateAssetData.ts [hunk 8]
```diff
@@ -90,7 +95,17 @@ const generateAssetData = async () => {
       .includes(asset.name)
   }
 
+  const currentGeneratedAssetData: AssetsByIdPartial = JSON.parse(
+    await fs.promises.readFile(generatedAssetsPath, 'utf8'),
```
Comment (NeOMakinG) [inline scripts/generateAssetData/generateAssetData.ts:99]: It remembers me something 👀

File: scripts/generateAssetData/generateRelatedAssetIndex/generateRelatedAssetIndex.ts [hunk 9]
```diff
@@ -229,11 +250,17 @@ export const generateRelatedAssetIndex = async () => {
...
+  )
+  const assetDataWithRelatedAssetKeys: Record<AssetId, PartialFields<Asset, 'relatedAssetKey'>> = {
+    ...generatedAssetData,
+  }
 
-  // remove relatedAssetKey from the existing data to ensure the related assets get updated
-  Object.values(assetDataWithRelatedAssetKeys).forEach(asset => delete asset.relatedAssetKey)
+  if (rebuildAll) {
+    // remove relatedAssetKey from the existing data to ensure the related assets get updated
+    Object.values(assetDataWithRelatedAssetKeys).forEach(asset => delete asset.relatedAssetKey)
```
Comment (NeOMakinG) [inline scripts/generateAssetData/generateRelatedAssetIndex/generateRelatedAssetIndex.ts:262]: Might be stupid but why if the `rebuildAll` prop is passed, we don't just nuke the file and recreate it from scratch?

**Review by NeOMakinG — COMMENTED** [review]
Conceptual review, I'm testing now!

**Review by NeOMakinG — APPROVED** [review]
The performance upgrade is lit!

I noticed `yarn generate:related-asset-index` produces a migration even if no changes has been done, don't you think that we could avoid this? Maybe by checksumming the file and check if it has been written or any other system!

Also, assets looks sane compared with develop, seems good for me!
image

image

image

File: scripts/generateAssetData/generateRelatedAssetIndex/generateRelatedAssetIndex.ts (hunk 9 above)
Comment (gomesalexandre) [reply scripts/generateAssetData/generateRelatedAssetIndex/generateRelatedAssetIndex.ts:262]: That's effectively what we do! Agree would be a bit more performant not to read the original file and assume empty one, but prefer not to add additional branching here 🙏🏽
